src/scratchOrgSettingsGenerator.ts

```diff
@@ -0,0 +1,421 @@
+        return false;
+      }
+
+      if (!(orgPrefApi in data)) {
+        set(data, orgPrefApi, {});
+      }
+      const apiOrgPrefs = ensureObject(getObject(data, orgPrefApi));
+
+      // check to see if the value is already set
+      set(apiOrgPrefs, mdApiName, prefVal);
+
+      return orgPrefApi !== OrgPrefRegistry.ORG_PREFERENCE_SETTINGS;
+    }
+
+    const orgPreferenceSettings = getObject(scratchDef, 'settings.orgPreferenceSettings') as Record<string, unknown>;
+    delete scratchDef.settings?.orgPreferenceSettings;
+    this.settingData = getObject(scratchDef, 'settings');
+
+    let migrated = false;
+    for (const preference in orgPreferenceSettings) {
+      if (storePrefs(ensureObject(this.settingData), preference, orgPreferenceSettings[preference])) {
+        migrated = true;
+      }
+    }
+
+    // Since we could have recommended some preferences that are still in OPS, only warn if any actually got moved there
+    if (migrated) {
+      // It would be nice if cli.ux.styledJSON could return a colorized JSON string instead of logging to stdout.
+      const message = messages.getMessage('migratedPrefFormat', [
+        oldScratchDef,
+        JSON.stringify({ settings: this.settingData }, null, 4),
+      ]);
+      warnings.push(message);
+    }
+    return warnings;
+  }
+
+  /** Create temporary deploy directory used to upload the scratch org shape.
+   * This will create the dir, all of the .setting files and minimal object files needed for objectSettings
+   */
+  public async createDeployDir(): Promise<string> {
+    // Base dir for deployment is always the os.tmpdir().
+    const shapeDirName = `shape_${Date.now()}`;
+    const destRoot = path.join(os.tmpdir(), shapeDirName);
+    const settingsDir = path.join(destRoot, 'settings');
+    const objectsDir = path.join(destRoot, 'objects');
+
+    try {
+      await fs.mkdir(settingsDir, { recursive: true });
+      await fs.mkdir(objectsDir, { recursive: true });
+    } catch (e) {
+      // If directory creation failed, the root dir probably doesn't exist, so we're fine
+      this.logger.debug('caught error:', e);
+    }
+
+    await Promise.all([this.writeSettingsIfNeeded(settingsDir), this.writeObjectSettingsIfNeeded(objectsDir)]);
+
+    // If SFDX_MDAPI_TEMP_DIR is set, copy settings to that dir for people to inspect.
+    const mdapiTmpDir = env.getString('SFDX_MDAPI_TEMP_DIR');
+    if (mdapiTmpDir) {
+      const tmpDir = path.join(mdapiTmpDir, shapeDirName);
+      this.logger.debug(`Copying settings to: ${tmpDir}`);
+      await fs.copyFile(destRoot, tmpDir);
+    }
+
+    return destRoot;
+  }
+
+  /**
+   * Deploys the settings to the org.
+   */
+  public async deploySettingsViaFolder(username: string, sourceFolder: string): Promise<FileResponse[]> {
+    this.logger.debug(`deploying settings from ${sourceFolder}`);
+
+    const componentSet = ComponentSet.fromSource(sourceFolder);
```

Comment:
of this entire PR, this is the one that's really causing me to think about options.  We're about to create a circular dependency SDR <=> sfdx-core.

It's not a bad one (SDR doesn't use this part of core, so it's not like 2 classes pointed at each other) but it might be a eventual operations pain (think version bumps and ts conflicts).  I'd love some others' thoughts about this.

The alternative would be returning a componentSet to the command(s), letting the command call deploy, and then having them maintain an instance or pass back the deploy results into sfdx-core.  All of which sounds nasty.

I think this might be the best option for sharing code with `sf`.